collectors/python.d.plugin/mail_log/mail_log.chart.py

```diff
@@ -0,0 +1,234 @@
+            return
+        service = match.group(2)
+        first = match.group(3)
+        if service == "qmgr":
+            self.parse_qmgr(line, first)
+        elif service == "smtpd" or service == "cleanup":
+            self.parse_smtpd_and_cleanup(line, first, service)
+        elif service == "smtp":
+            self.parse_smtp(line)
+
+
+class Service(LogService):
+    def __init__(self, configuration=None, name=None):
+        LogService.__init__(self, configuration=configuration, name=name)
+        self.order = ORDER
+        self.definitions = CHARTS
+        self.stats = MailStatistics()
+        self.parser = {
+            "postfix": PostfixParser
+        }.get(configuration.get("type", "postfix"))(self)
+
+    def check(self):
+        """
+        :return: bool
+
+        1. "log_path" is specified in the module configuration file
+        2. "log_path" must must exist and be readable by netdata user
+        """
+        if not self.log_path:
+            self.error('log path is not specified')
+            return False
+
+        if not (self._find_recent_log_file() and os.access(self.log_path, os.R_OK)):
+            self.error('{log_file} not readable or does not exist'.format(log_file=self.log_path))
+            return False
+        return True
+
+    def parse_line(self, line):
+        self.parser.parse_line(line)
+
+    def _get_data(self):
+        """
+        Parses new log lines
+        :return: dict OR None
+        None if _get_raw_data method fails or parsing error encountered.
+        In all other cases - dict.
+        """
+        try:
+            for line in self._get_raw_data():
+                self.parse_line(line)
+            data = {
+                "mail_connections": self.stats.connections,
+                "mail_accepted": self.stats.accepted,
+                "mail_greylisted": self.stats.greylisted,
+                "mail_temp_rejected": self.stats.temp_rejected,
+                "mail_perm_rejected": self.stats.perm_rejected,
+                "mail_sent": self.stats.sent,
+                "mail_deferred": self.stats.deferred,
+                "mail_bounced": self.stats.bounced
+            }
+
+            for code, count in self.stats.incoming_codes.items():
+                data["mail_incoming_codes_" + code] = count
+            for code, count in self.stats.outgoing_codes.items():
+                data["mail_outgoing_codes_" + code] = count
+
+            return data
+        except (ValueError, AttributeError) as ex:
+            self.error(ex)
+            return None
+
+    def increment_incoming_code(self, code):
+        """Increments number of incoming messages per status code, adds a new line to chart if required"""
+        if code not in self.stats.incoming_codes:
+            self.charts["incoming_codes"].add_dimension(['mail_incoming_codes_' + code, code, 'incremental', 1, 1])
+            self.stats.incoming_codes[code] = 0
+        self.stats.incoming_codes[code] += 1
```

Comment:
This should be `update_incoming_codes_chart` and `update_outgoing_codes_chart` methods and this should be handles by `Service`, not parser.